IDSEQ-1552 - Implement Auth0 Alpha Login Strategy

davidrissato opened the pull request:
# Description

Add auth0 login strategy in parallel to Devise.

We are currently using devise to handle authentication and authorization. As an intermediary migration step, add support to auth0 login strategy without removing Devise legacy authentication.

# Tests

* Manual tests using auth0
* Manual tests using previous authentication method

davidrissato commented:
Looks like I still need to fix something here

davidrissato commented:
Let me close this PR while I fix the issue. I'll reopen it soon.

jshoe commented on app/assets/src/components/utils/links.js:
```diff
@@ -43,10 +43,31 @@ const openUrlInPopupWindow = (url, windowName, windowWidth, windowHeight) => {
   );
 };
 
+const postToUrlWithCSRF = (url, params) => {
```
this function might be better in `src/api/core.js`

davidrissato replied:
I took a look on `core.js`, and it seems the functions over there are all used to invoke json endpoints, whereas `links.js` has functions that replace the current location (which is the case here, with the difference I'm using a post operation to do that).
That said, it looked to me to be more logical to put this method on `links.js`. What do you think?

jshoe replied:
I see, `links.js` sounds good then

jshoe commented on app/controllers/application_controller.rb:
```diff
@@ -27,6 +24,34 @@ def admin_required
   end
 
+  # This method is used to verify if the user is already logged in,
+  # and will redirect it to the homepage in case it is not.
```
nit: "in case they are not"

jshoe commented on app/controllers/application_controller.rb:
```diff
@@ -27,6 +24,34 @@ def admin_required
+  # This method is used to verify if the user is already logged in,
+  # and will redirect it to the homepage in case it is not.
+  # This method overrides a default behavior from Devise to allow auth0
+  # authentication working simultaneously with legacy devise database mode,
```
nit: "to work simultaneously"

jshoe commented on app/controllers/application_controller.rb:
```diff
@@ -27,6 +24,34 @@ def admin_required
+        return if current_user&.email == auth_payload["email"]
+        auth_user = User.find_by(email: auth_payload["email"])
+        if auth_user.present?
+          sign_in auth_user, store: false
```
what does `store: false` do?

davidrissato replied:
Devise documentation doesn't have info about that, but from what I could check, it prevents warden from creating a session, which is exactly what I wanted since I'm handling the session in a different place.
I could find some extra info about this behavior [here](https://github.com/plataformatec/devise/issues/3981)

jshoe commented on app/controllers/application_controller.rb:
```diff
@@ -81,26 +106,28 @@ def get_background_id(sample)
-
-  # Modified from https://gist.github.com/josevalim/fb706b1e933ef01e4fb6
-  def authenticate_user_from_token!
+  # This method is only used to loging users using tokens
```
sp: s/loging/login

jshoe commented on app/controllers/application_controller.rb:
```diff
@@ -81,26 +106,28 @@ def get_background_id(sample)
+  # It is not intended to replace the authenticate_user! method, which
+  # must be invoked regardless.
+  # authenticate_user! is used for verifying if the user is already logged in
+  # and redirecting to the homepage in case it is not.
```
nit: "in case they are not"

jshoe commented on app/controllers/auth0_controller.rb:
```diff
@@ -0,0 +1,16 @@
+    self.auth0_session = request.env['omniauth.auth'].credentials.id_token
+
+    # Redirect to the URL you want after successful auth
+    redirect_to "/home"
```
can you do `redirect_to root_path` instead

davidrissato replied:
I noticed `root_path == "/"`. I wanted to redirect to home, since user is already authenticated. Let me see if there is a more elegant way to do that.

davidrissato replied:
Ok, looks like `home_path` is what I was looking for

jshoe commented on app/helpers/auth0_helper.rb:
```diff
@@ -0,0 +1,62 @@
+  protected
+
+  def auth0_signout_url
+    domain = Rails.application.secrets.auth0_domain
```
should we also get these from `ENV["AUTH0_DOMAIN"]` to be consistent?

davidrissato replied:
Actually I should be invoking this method instead. I started moving it here but it seems I ended up not completely doing that. Thank you for catching this.

jshoe commented on app/lib/json_web_token.rb:
```diff
@@ -0,0 +1,43 @@
+    end
+  end
+
+  def self.jwks_hash
```
could you add a comment or two to these methods about their function for future readers?

davidrissato replied:
Done

jshoe commented on app/views/home/landing.html.erb:
```diff
@@ -7,7 +7,8 @@
         showBulletin: <%= @show_bulletin %>,
-        browserInfo: <%= @browser_info.to_json.html_safe %>
+        browserInfo: <%= @browser_info.to_json.html_safe %>,
+        signInStrategy: location.href.match(/auth0/) ? 'AUTH0' : 'DEVISE',
```
nit: could you move this line into the `HomeController#landing` method like `@sign_in_strategy`? Just want to keep these erb files very slim on logic

jshoe commented on bin/entrypoint.sh:
```diff
@@ -1,9 +1,16 @@
     exec bundle exec "$@"
 else
     # Use Chamber to inject secrets via environment variables.
+    echo Chamber will reading entries from AWS SSM /idseq-$ENVIRONMENT-web/
```
nit: s/reading/read

davidrissato commented:
Found an issue in this PR. I'm closing it while I fix it, and I'll reopen it later on.

davidrissato commented:
Reopening this PR since. There was only a small issue related to not verifying the audience.
I also added comments to `json_web_token.rb` and added a caching mechanism for `jwks_hash`
